enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,244 @@
+2. OLM calculates the latest generation of operators that should be installed in a namespace and creates an InstallPlan listing each resource that must exist on the cluster.
+3. If the InstallPlan requires manual approval, a User must "approve" it.
+5. OLM creates/updates the manifests defined in the InstallPlan
+6. If an older CSV exists on the cluster OLM will place in the "Replacing" phase.
+7a. OLM fails to create/update a resource defined in the InstallPlan and moves the InstallPlan to the "Failed" state.
+7b. OLM installs a CSV and it enters the "Failed" state or forever enters the "Pending" state.
+
+In either failure scenario, a cluster admin must perform manual steps to recover, which becomes increasingly untenable as the number of managed clusters increases.
+
+> Note: OLM does not perform multiple attempts to resolve InstallPlans, meaning that a failed install blocks upgrades for all operators in the namespace as resolution is on a per-namespace basis.
+
+This enhancement proposes allowing cluster admins to opt-into fail forward upgrades, thereby enabling OLM to recover from failed installations when a new upgrade path is discovered.
+
+### User Stories
+
+- As a cluster admin, I want to allow all operators to "fail forward" in a specific namespace.
+- As a cluster admin, if an InstallPlan is in the FAILED state, I want to create the next InstallPlan.
+- As a cluster admin, if a CSV is stuck in the PENDING state, I want to allow it to upgrade to the next version.
+- As a cluster admin, if a CSV is stuck in the FAILED state, I want to allow it to upgrade to the next version.
+- As a catalog curator / operator author, I want to chose which versions of an operator may "fail forward" to the next version so my customers don't miss critical upgrades.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Opting into Fail Forward Upgrades
+
+Since resolution is namespaced, the toggle for allowing forced upgrades also should be namespaced to avoid having to add the extra burden on the resolver surrounding understanding which dependent operators can or cannot use the forced upgrade toggle.
+
+With this in mind, this enhancement proposes adding the `failForwardUpgrades` toggle to an already watched namespaced resource, the OperatorGroup.
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
+metadata:
+  name: foo
+  namespace: bar
+spec:
+  failForwardUpgrades: true
+```
+
+With the toggle enabled, OLM will allow operators to "fail forward" if such upgrades are supported by the operator. If the toggle is disabled, OLM exhibits existing behavior.
+
+#### Supporting Fail Forward Upgrades
+
+This enhancement proposes the introduction of the `fail-forward-supported` [Bundle Property](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#properties). When the resolver determines the set of operators to install, OLM can place constraints on a bundle that must be met in the next generation of installables.
```

Comment:
Having the fail-forward toggle be a property of a particular bundle makes sense, but it puts a lot of the complexity of reasoning about failed installation upgrades to the maintainer of the particular package in which the bundle is placed. Would it be valuable to have a way to have the fail-forward property be a property of the entire package or entire channel as well? 

I'm not sure how well it works with what the resolver is expecting, but you could see a scenario where being able to say "all upgrades within this alpha channel should fail forward" is very useful. Maybe this is a followup to the original implementation.

Reply:
We already have bundle level properties editable on FBC which the resolver can act on easily as of now. We could enable a channel or package-wide toggle instead depending on how tightly we want to control failing forward. The cluster admins would likely still need to update the catalog to have the new upgrade edge so it could be just a matter of ensuring the fail-forward property when they do.

Currently the resolver doesn't look at package or channel level properties and we don't have them returned via any of the existing GRPC endpoints from the registry. FBC does have support for them though so it is worth discussing.